tests/ml/test_ml_model_pytest.py

```diff
@@ -306,6 +306,68 @@ def test_decision_tree_regressor(self, compress_model_definition):
         # Clean up
         es_model.delete_model()
 
+    @requires_sklearn
+    @pytest.mark.parametrize("compress_model_definition", [True, False])
+    def test_learning_to_rank(self, compress_model_definition):
+        # Train model
+        training_data = datasets.make_regression(n_features=2)
+        regressor = DecisionTreeRegressor()
+        regressor.fit(training_data[0], training_data[1])
+
+        # Serialise the models to Elasticsearch
+        feature_names = [f"feature_{i}" for i in range(2)]
+        model_id = "test_learning_to_rank"
+        inference_config = {
+            "learning_to_rank": {
+                "feature_extractors": [
+                    {
+                        "query_extractor": {
+                            "feature_name": "feature_1",
+                            "query": {
+                                "script_score": {
+                                    "query": {"match_all": {}},
+                                    "script": {"source": 'return doc["price"].value;'},
+                                }
+                            },
+                        }
+                    },
+                    {
+                        "query_extractor": {
+                            "feature_name": "feature_2",
+                            "query": {"term": {"type": "TV"}},
+                        }
+                    },
+                ]
+            }
+        }
+
+        es_model = MLModel.import_model(
+            ES_TEST_CLIENT,
+            model_id,
+            regressor,
+            feature_names,
+            es_if_exists="replace",
+            es_compress_model_definition=compress_model_definition,
+            inference_config=inference_config,
+        )
+
+        # Verify the saved inference config contains the passed LTR config
+        response = ES_TEST_CLIENT.ml.get_trained_models(model_id=model_id)
+        assert response.meta.status == 200
+        assert response.body["count"] == 1
+        saved_inference_config = response.body["trained_model_configs"][0][
+            "inference_config"
+        ]
+        assert "learning_to_rank" in saved_inference_config
+        saved_ltr_config = saved_inference_config["learning_to_rank"]
+        assert all(
+            item in saved_ltr_config.items()
+            for item in inference_config["learning_to_rank"].items()
+        )
+
```

Comment:
You should be able to use the model as a learning to rank rescorer on the movie index:

```
ES_TEST_CLIENT.search(
  index = MOVIES_INDEX_NAME,
  query = { 
    "multi_match": { 
      "fields": ["title", "actors", "directors" , "plot"], 
      "query": "planet of the apes" 
    }
  },
  rescore = { 
    "learning_to_rank": { 
      "model_id": model_id, 
      "params": {"query_string": "planet of the apes" } 
    } 
  }
)
```

Reply:
Potentially, the scores extracted by the feature extractors are too close and our completely randomly trained model returns the same score for them.

I did try a completely different test on my env:

```python
# Execute search with rescoring and verify document order
search_response = ES_TEST_CLIENT.search(
    index=MOVIES_INDEX_NAME,
    query={"terms": {"_id": ["tt1318514", "tt0071562"] } },
    rescore={
        "learning_to_rank": {
            "model_id": model_id,
            "params": {"query_string": "planet of the apes"},
        }
    },
)

doc_scores = [hit['_score'] for hit in search_response['hits']['hits']]

# Assert that:
# - all documents from the query are present
# - all documents have been rescored (score != 1.0)
# - document scores are unique
assert len(search_response['hits']['hits']) == 2
assert all(score != float(1) for score in doc_scores)
assert all(doc_scores.count(score) == 1 for score in doc_scores)
```

I rescore two documents that are very different so it is unlikely that they get close scores during feature extraction.

Also, it seems to me a little bit trappy to assume our random regression model will rescrore to preserve an order (it could do the inverse). Instead of doc ordering, I prefer to test that all docs have been rescored and that all scores are different.

WDYT?

Reply:
I support the idea because testing the rescorer's behavior is out of scope for the unit test. So as long as we verify it does something to the scores, we're good.
I'll make the code changes as suggested.